hie-plugin-api/Haskell/Ide/Engine/Ghc.hs

```diff
@@ -130,131 +129,185 @@ srcErrToDiag df rfm se = do
+    setTypecheckedModule_load uri
+
+-- Hacky, need to copy hs-boot file if one exists for a module
+-- This is because the virtual file gets created at VFS-1234.hs and
+-- then GHC looks for the boot file at VFS-1234.hs-boot
+--
+-- This strategy doesn't work if the user wants to edit the boot file but
+-- not save it and expect the VFS to save them. However, I expect that HIE
+-- already didn't deal with boot files correctly.
+copyHsBoot :: FilePath -> FilePath -> IO ()
+copyHsBoot fp mapped_fp = do
+  ex <- doesFileExist (fp <> "-boot")
+  if ex
+    then copyFile (fp <> "-boot") (mapped_fp <> "-boot")
+    else return ()
+
+loadFile :: (FilePath -> FilePath) -> (FilePath, FilePath)
+         -> IdeGhcM (Diagnostics, AdditionalErrs,
+                     Maybe (Maybe TypecheckedModule, [TypecheckedModule]))
+loadFile rfm t = do
+    withProgress "loading" NotCancellable $ \f -> (captureDiagnostics rfm $ BIOS.loadFileWithMessage (Just $ toMessager f) t)
+
+-- | Actually load the module if it's not in the cache
+setTypecheckedModule_load :: Uri -> IdeGhcM (IdeResult (Diagnostics, AdditionalErrs))
+setTypecheckedModule_load uri =
   pluginGetFile "setTypecheckedModule: " uri $ \fp -> do
-    fileMap <- GM.getMMappedFiles
-    debugm $ "setTypecheckedModule: file mapping state is: " ++ show fileMap
-    rfm <- GM.mkRevRedirMapFunc
-    let
-      ghcErrRes msg = ((Diagnostics Map.empty, [T.pack msg]),Nothing,Nothing)
-      progTitle = "Typechecking " <> T.pack (takeFileName fp)
     debugm "setTypecheckedModule: before ghc-mod"
+    debugm "Loading file"
+    mapped_fp <- persistVirtualFile uri
+    liftIO $ copyHsBoot fp mapped_fp
+    rfm <- reverseFileMap
     -- TODO:AZ: loading this one module may/should trigger loads of any
     -- other modules which currently have a VFS entry.  Need to make
     -- sure that their diagnostics are reported, and their module
     -- cache entries are updated.
     -- TODO: Are there any hooks we can use to report back on the progress?
-    ((Diagnostics diags', errs), mtm, mpm) <- withIndefiniteProgress progTitle NotCancellable $ GM.gcatches
-      (GM.getModulesGhc' (myWrapper rfm) fp)
-      (errorHandlers ghcErrRes (return . ghcErrRes . show))
-    debugm "setTypecheckedModule: after ghc-mod"
-
+    (Diagnostics diags', errs, mmods) <- loadFile rfm (fp, mapped_fp)
+    debugm "File, loaded"
     canonUri <- toNormalizedUri <$> canonicalizeUri uri
     let diags = Map.insertWith Set.union canonUri Set.empty diags'
-    diags2 <- case (mpm,mtm) of
-      (Just pm, Nothing) -> do
-        debugm $ "setTypecheckedModule: Did get parsed module for: " ++ show fp
-        cacheModule fp (Left pm)
-        debugm "setTypecheckedModule: done"
-        return diags
-
-      (_, Just tm) -> do
+    debugm "setTypecheckedModule: after ghc-mod"
+    debugm ("Diags: " <> show diags')
+    let collapse Nothing = (Nothing, [])
+        collapse (Just (n, xs)) = (n, xs)
+
+    case collapse mmods of
```

Comment:
We don't save the parsed module here on its own. We discussed this quite a bit and decided that saving the parsed modules was of dubious benefit.